### PR TITLE
[UTOPIA-1286] allow user to complete PIA

### PR DESCRIPTION
**Pull request by binggaofw**
# Description

This PR includes the following proposed change(s):

- see title
Screenshot 2023-07-14 at 3 03 32 PM

## Type of change

- [x] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Refactoring / Documentation
- [ ] Updating Testing Framework(s)
- [ ] Version change

## How Has This Been Tested?

Please describe the tests that you ran to verify your changes.

## Development Dependency Working Agreement
- [x] My code DOES NOT include the importing of new dependencies into the DPIA ecosystem
- [ ] My code DOES include the importing of new dependencies into the DPIA ecosystem
**If new dependencies are being introduced to the DPIA ecosystem:**
- [ ] The functionality of the dependency drastically reduces code complexity and makes my changes more easily maintainable and readible 
- [ ] The dependency being introduced does not contain multiple layers of nested dependencies introducing maintainability complexity to the DPIA ecosystem

## Frontend Development Changes
- [ ] N/A
- [x] I have performed a self-review of my own code
- [ ] I have made corresponding changes to project documentation or diagrams that reflect my changes
- [ ] New and existing unit tests pass locally with my changes
- [ ] My code follows Airbnb React Style Guidelines

## API Development Changes
- [ ] N/A
- [x] I have performed a self-review of my own code
- [ ] My code follows standards and practices outlined in the BC Government API Development Guidelines
- [ ] New and existing unit tests pass locally with my changes
- [ ] My changes includes Swagger documentation updates that reflect the changes I am introducing

## Definition of Done

Definition of Done

### CODE REVIEW

**Comment by binggaofw** [comment]
this  change only for front end, the backend need to update as well, @kushal-arora-fw  please create a backend pr  thanks

File: src/frontend/src/components/public/PIASubHeader/index.tsx [hunk 1]
```diff
@@ -120,6 +121,37 @@ function PIASubHeader({
...
+      const selectedRoles = pia?.review?.programArea?.selectedRoles || [];
+      for (const role of selectedRoles) {
+        if (
+          pia?.review?.programArea?.reviews?.[role]?.isAcknowledged !== true
+        ) {
+          reviewProgramAreaDone = false;
+          break;
+        }
+        reviewProgramAreaDone = true;
+      }
```
Comment (kushal-arora-fw) [inline src/frontend/src/components/public/PIASubHeader/index.tsx:137]:
```suggestion
const reviewProgramAreaDone = selectedRoles.every((role) => pia?.review?.programArea?.reviews?.[role]?.isAcknowledged === true)
```

File: src/frontend/src/pages/PIAForm/index.tsx [hunk 2]
```diff
@@ -610,6 +617,20 @@ const PIAFormPage = () => {
...
+        if (updatedPia?.id) {
+          navigate(
+            buildDynamicPath(routes.PIA_VIEW, {
+              id: updatedPia.id,
+            }),
+          );
+        }
+      } else if (buttonValue === 'completePIA') {
+        const updatedPia = await upsertAndUpdatePia({
+          // here not sure what status for this one, need to discuss
```
Comment (kushal-arora-fw) [inline src/frontend/src/pages/PIAForm/index.tsx:629]: do you need discussion here?

File: src/frontend/src/utils/status.ts [hunk 3]
```diff
@@ -362,6 +362,16 @@ export const statusList = (pia: IPiaForm | null): StatusList => {
                 cancelLabel: 'Cancel',
               },
             },
+            {
+              status: 'COMPLETE',
+              modal: {
+                title: 'Submit for Completion',
+                description:
+                  'Once CPO has confirmed all necessary ministry reviews have occurred and data has been uploaded to the PID, PIA will move to “Complete” status.',
```
Comment (kushal-arora-fw) [inline src/frontend/src/utils/status.ts:370]: Is the content same for Delegated flow as well, where CPO involvement is not there at all?

File: src/frontend/src/components/public/PIASubHeader/index.tsx [hunk 4]
```diff
@@ -120,6 +121,37 @@ function PIASubHeader({
...
+    }
+
+    if (
```
Comment (antsand) [inline src/frontend/src/components/public/PIASubHeader/index.tsx:132]: please add this in a separate function with proper commenting.
If  all reviews are acknnowledged
the currest status is in final review and
MPO too has acknowledged the PIA can progress to complete.

File: src/frontend/src/components/public/PIASubHeader/index.tsx [hunk 5]
```diff
@@ -133,10 +157,21 @@ function PIASubHeader({
       primaryButtonText === SubmitButtonTextEnum.DELEGATE_FINISH_REVIEW
```
Comment (antsand) [inline src/frontend/src/components/public/PIASubHeader/index.tsx:157]: this logic is strange. Why can't we check if pia is delegated and the status is in complete review?

File: src/frontend/src/components/public/PIASubHeader/index.tsx (hunk 4 above)
Comment (binggaofw) [reply src/frontend/src/components/public/PIASubHeader/index.tsx:132]: the useEffect is kind of a function, but I will add more comments

File: src/frontend/src/utils/status.ts (hunk 3 above)
Comment (binggaofw) [reply src/frontend/src/utils/status.ts:370]: confirm with UX just apply for delegate PIA, I think I just remove this from CPO list

File: src/frontend/src/pages/PIAForm/index.tsx (hunk 2 above)
Comment (binggaofw) [reply src/frontend/src/pages/PIAForm/index.tsx:656]: confirm, right now route to pia intake tab

File: src/frontend/src/components/public/PIASubHeader/index.tsx (hunk 5 above)
Comment (binggaofw) [reply src/frontend/src/components/public/PIASubHeader/index.tsx:163]: i think both implementation are the same.

File: src/frontend/src/components/public/PIASubHeader/index.tsx [hunk 6]
```diff
@@ -133,22 +163,38 @@ function PIASubHeader({
...
       return false;
+
     return true;
   };
   const showSubmitButton = () => {
     const owner = getGUID() === pia.createdByGuid ? true : false;
-    // if the status is final review, hide submit button
-    if (pia.status === PiaStatuses.FINAL_REVIEW) return false;
+    // if the status is completed, hide submit button
+    if (pia.status === PiaStatuses.COMPLETE) return false;
```
Comment (antsand) [inline src/frontend/src/components/public/PIASubHeader/index.tsx:197]: Can we add this logic to the state table. IN the complete staus have a parameter submit button and do the evaluation there?

File: src/frontend/src/pages/PIAForm/index.tsx [hunk 7]
```diff
@@ -266,7 +267,9 @@ const PIAFormPage = () => {
     const onNewPiaPage = pathname === buildDynamicPath(routes.PIA_NEW, {});
 
     // if the user is on intake or new PIA page, show the submit PIA intake button; Else submit PIA
-    if (
+    if (pia.status === PiaStatuses.FINAL_REVIEW) {
```
Comment (antsand) [inline src/frontend/src/pages/PIAForm/index.tsx:270]: Similarly, please keep anything related to status logic within the status state table. We should NOT be having this type of checks all over the place.

**Review by kushal-arora-fw — DISMISSED** [review]

**Review by antsand — APPROVED** [review]

**Pull request merged**